niriPipe/utils/finder.py

```diff
@@ -0,0 +1,396 @@
+        found.
+        """
+        self.logger.debug("Finding {} frames.".format(frame_type))
+        key = 'min_{}s'.format(frame_type)
+        try:
+            table = self._do_query(query)
+        except Exception as e:
+            if self.state['config']['DATAFINDER'][key]:
+                logging.critical("{} query failed.".format(frame_type))
+                raise e
+            else:
+                self.logger.debug("Found no frames; returning empty table.")
+                return astropy.table.Table(names=('publisherID', 'productID'))
+
+        if len(table) < \
+                self.state['config']['DATAFINDER'][key]:
+            raise RuntimeError("Required {} {} frames; found {}.".format(
+                self.state['config']['DATAFINDER'][key],
+                frame_type,
+                len(table)))
+        self.logger.debug("Found {} {} frames.".format(len(table), frame_type))
+        return table
+
+    def _do_query(self, query):  # pragma: no cover
+        """
+        Does a CADC async query.
+        """
+        cadc = Cadc()
+        job = cadc.create_async(query)
+        job.run().wait()
+        job.raise_if_error()
+        return job.fetch_result().to_table()
+
+    def _metadata_from_header(self, productID, card):  # pragma: no cover
+        """
+        Use cadc-data to get metadata not findable by tap.
+
+        This method is a hack to get around metadata not available in CADC TAP.
+        The better solution is to get all metadata possible from CADC TAP.
+        There are probably better ways to do this, but the intention is to
+        minimize the use of this method.
+        """
+        anonSubject = net.Subject()
+        client = CadcDataClient(anonSubject)
+        with io.BytesIO() as f:
+            f.name = None
+            client.get_file('GEM', productID, f, fhead=True)
+            contents = f.getvalue().decode('utf-8')
+            if card == 'CAMERA':
+                pattern = re.compile(r'(?<=CAMERA  \= \')[^\s]*')
+                return pattern.search(contents).group()
+            else:
+                raise ValueError("Card {} not implemented.".format(card))
+
+    def _segment(self, in_table):
+        """
+        Narrow a table down to the single observation closest to the object
+        observation in time.
+
+        First, compute an observation name column for the table. Then, for
+        each observation name, calculate the time difference between it
+        and the object frames (use an average here of the first and last
+        frames in the observation). Return the frames of that closest
+        observation.
+
+        The failure of segmentation might cause downstream errors, so log
+        a warning if it fails.
+        """
+        self.logger.debug(
+            "Segmentation starting with {} frames.".format(len(in_table)))
+        # Do segmentation based on state['current_stack']['mjd+date']
+        in_table.sort(['publisherID'])
```

Comment:
An `ORDER BY publisherID ASC` or `ORDER BY publisherID DESC` clause in the query text would do this as well. This would off-load cpu cycles to the database server.

Reply:
Done.